File: packages/types/src/risk.ts
export const DECISIONS = ["ALLOW", "CHALLENGE", "BLOCK", "REVIEW"] as const;
export type Decision = (typeof DECISIONS)[number];
export type RiskLevel = "low" | "medium" | "high" | "critical";
export type EvaluationMode = "Observe" | "Enforce";

export interface Evidence {
  event: string;
  weight?: number;
  seconds_after_previous?: number;
  amount_vs_customer_median?: number;
  balance_usage_ratio?: number;
  observed?: string;
  interpretation?: string;
}

export interface RiskEvaluation {
  evaluation_id: string;
  decision: Decision;
  risk_score: number;
  risk_level: RiskLevel;
  recommended_action: Lowercase<Decision>;
  policy_rule: string;
  primary_reason: string;
  evidence: Evidence[];
  model_version: string;
  created_at: string;
  institution_ref?: string | null;
  mode: EvaluationMode;
  enforced: boolean;
  idempotent_replay: boolean;
  latency_ms?: number | null;
  fallback?: boolean;
}

File: apps/api/nche/schemas/risk.py
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field

from .events import NcheEvent


class Decision(StrEnum):
    ALLOW = "ALLOW"
    CHALLENGE = "CHALLENGE"
    BLOCK = "BLOCK"
    REVIEW = "REVIEW"


class EvaluationMode(StrEnum):
    """Controls whether the recommendation is shadowed or enforced."""

    OBSERVE = "Observe"
    ENFORCE = "Enforce"


class RiskLevel(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class Evidence(BaseModel):
    event: str
    weight: float | None = None
    seconds_after_previous: int | None = None
    amount_vs_customer_median: float | None = None
    balance_usage_ratio: float | None = None
    observed: str | None = None
    interpretation: str | None = None


class RiskRequest(BaseModel):
    customer_ref: str
    events: list[NcheEvent] = Field(min_length=1)
    institution_ref: str | None = None


class RiskEvaluation(BaseModel):
    evaluation_id: str
    decision: Decision
    risk_score: int = Field(ge=0, le=100)
    risk_level: RiskLevel
    recommended_action: str
    policy_rule: str
    primary_reason: str
    evidence: list[Evidence]
    model_version: str
    created_at: datetime
    institution_ref: str | None = None
    mode: EvaluationMode = EvaluationMode.OBSERVE
    enforced: bool = False
    idempotent_replay: bool = False
    latency_ms: float | None = Field(default=None, ge=0)
    fallback: bool = False
